### PR TITLE
Add function to convert node tree back to HTML

### PR DESCRIPTION
**Pull request by glaucocustodio**

### CODE REVIEW

File: lib/floki.ex [hunk 1]
```diff
@@ -77,6 +77,50 @@ defmodule Floki do
...
+  def raw_html([], html), do: html
+  def raw_html([value|tail], html) when is_bitstring(value), do: value
+  def raw_html([first_dom|tail], html \\ "") do    
+    elem  = first_dom |> elem(0)
+    attrs = first_dom |> elem(1) |> tag_attrs
+    value = first_dom |> elem(2)
+    raw_html(tail, html <> tag_for(elem, attrs, value))
+  end
+
+  def tag_attrs(attr_list) do
```
Comment (philss) [inline lib/floki.ex]: I think this function could be private.

File: lib/floki.ex [hunk 2]
```diff
@@ -77,6 +77,50 @@ defmodule Floki do
...
+    attr_list
+    |> Enum.reduce("", fn(c,t) -> ~s(#{t} #{elem(c,0)}="#{elem(c,1)}") end)
+    |> String.strip
+  end
+
+  def tag_for(elem, attrs, value) do
```
Comment (philss) [inline lib/floki.ex]: This function could be private as well.

File: lib/floki.ex [hunk 3]
```diff
@@ -77,6 +77,50 @@ defmodule Floki do
...
+      iex> Floki.parse(~s(<div class="wrapper">my content</div>)) |> Floki.raw_html
+      ~s(<div class="wrapper">my content</div>)
+
+  """
+
+  def raw_html(tuple) when is_tuple(tuple), do: raw_html([tuple])
+  def raw_html([], html), do: html
+  def raw_html([value|tail], html) when is_bitstring(value), do: value
+  def raw_html([first_dom|tail], html \\ "") do    
+    elem  = first_dom |> elem(0)
```
Comment (philss) [inline lib/floki.ex]: You can use pattern matching to get the values, like

```elixir
{elem, attrs, value} = first_dom
```

File: test/floki_test.exs [hunk 4]
```diff
@@ -119,6 +132,41 @@ defmodule FlokiTest do
...
+    raw_html = Floki.parse(@html_with_data_attributes) |> Floki.raw_html
+    assert raw_html == String.split(raw_html, "\n") |> Enum.map(&(String.strip(&1))) |> Enum.join("")
+  end
+
+  test "raw_html (after find)" do
+    raw_html = Floki.parse(@basic_html) |> Floki.find("a") |> Floki.raw_html
+    assert raw_html == ~s(<a href="uol.com.br" class="bar"><span>UOL</span><img src="foo.png"/></a>)
+  end
+
+  # Floki.tag_attrs/1
```
Comment (philss) [inline test/floki_test.exs]: As I suggested to change those functions to private, you can remove the tests below.

**Comment by philss** [comment]
@glaucocustodio

I added some comments in the PR. Please check.
Also, I think you could change the variable name of `first_dom` to `node`, in order to keep consistency.

This is a very nice feature, thanks! :smiley:

**Comment by glaucocustodio** [comment]
Nice. I thought about keeping these functions you mentioned as private, but I didn't do in order to test them..
Anyway, I will do the requested changes. Thanks for the tips!

**Pull request merged**